Spark solr nested df

ganeshk7 opened the pull request.

thelabdude commented:
looks good overall, but can we think of a better name for SolrRDD.readDataFrame that indicates that it works with these nested structures?

ganeshk7 commented:
Thanks for the review Tim. I thought of adding couple of more things to this. 

1) Make use of the config map in SolrRelation class to let the user decide whether the schema needs to be preserved or not (May be a parameter named "preserveSchema"). If yes, then go ahead and store the way we store nested structures. If not, it will work the way it is right now. [By default if the "preserveSchema" is yes, it will preserve the schema, nester structures and the order of the dataframe columns]
2) If I understand correctly, "buildScan" method in SolrRelation class is called during the dataFrameRead? If yes, try and see if schema is there in Solr and call appropriate code if it is there. If not, it will work the way it is right now.

Please let me know your thoughts.

matthoffman commented on src/main/java/com/lucidworks/spark/SolrRelation.java:
```diff
@@ -63,8 +74,17 @@ public SolrRelation(SQLContext sqlContext, Map<String,String> config, DataFrame
-    } else {
-      schema = solrRDD.getQuerySchema(solrQuery);
+    }
+    else {
```
nitpick, but go ahead and put the else on the same line as the bracket, like `} else {`.  Just to be consistent style-wise.

matthoffman commented on src/main/java/com/lucidworks/spark/SolrRelation.java:
```diff
@@ -63,8 +74,17 @@ public SolrRelation(SQLContext sqlContext, Map<String,String> config, DataFrame
-      schema = solrRDD.getQuerySchema(solrQuery);
+    }
+    else {
+      if(!preserveSchema.equals("Y")) {
```
what's the convention for other Spark properties for indicating "true"?  Do they pass in "Y/N", or "true/false"?

matthoffman commented on src/main/java/com/lucidworks/spark/SolrRelation.java:
```diff
@@ -31,13 +39,15 @@
   public static String SOLR_SPLITS_PER_SHARD_PARAM = "splits_per_shard";
-
+  public static String PRESERVE_SCHEMA = "preserveschema";
+  protected String preserveSchema;
```
let's make this a boolean, so we can parse the property into a true/false in just one place and then just use normal boolean comparisons on it everywhere else. Seems safer.

matthoffman commented on src/main/java/com/lucidworks/spark/SolrRelation.java:
```diff
@@ -99,39 +119,61 @@ public StructType schema() {
+      try {
+        // build the schema based on the desired fields
+        StructType querySchema = (fields != null && fields.length > 0) ? deriveQuerySchema(fields) : schema;
+        JavaSparkContext jsc = new JavaSparkContext(sqlContext.sparkContext());
```
JavaSparkContext is available as a field now, right?

matthoffman commented on src/main/java/com/lucidworks/spark/SolrRelation.java:
```diff
@@ -223,23 +265,566 @@ protected String fq(Filter f) {
+      docs = convertToSolrDocuments(df, new HashMap<String,Object>());
+    }
+
+    SolrSupport.indexDocs(solrRDD.zkHost, solrRDD.collection, 100, docs);
```
should this be "sendToSolr(docs)" instead?

ganeshk7 replied:
sendToSolr(docs) was carried over from the previous implementation. The insert method already has the line "SolrSupport.indexDocs(solrRDD.zkHost, solrRDD.collection, 100, docs)" to send the docs to solr. I will remove the sendToSolr method.

ganeshk7 commented:
@matthoffman : Thanks for the review suggestions. I have changed the code as per the suggestion. Please review.

thelabdude commented:
Went back through these changes this morning as we need to do a release of this project soon. I'm now thinking it makes sense to just sub-class SolrRDD instead of doing this inline. That way, your code can just use the subclass directly vs. this preserve schema flag. You can add an option in SolrRelation to load your subclass instead of SolrRDD

ganeshk7 commented:
Sure Tim. Will get this done as soon as possible

ganeshk7 commented:
@matthoffman @thelabdude : The suggested changes has been done. Please review. Thanks

thelabdude commented on src/main/java/com/lucidworks/spark/SolrRelation.java:
```diff
@@ -223,23 +242,178 @@ protected String fq(Filter f) {
+  }
+
+
+  public JavaRDD<SolrInputDocument> convertToSolrDocuments(DataFrame df, final HashMap<String,Object> uniqueIdentifier) {
```
I'd move this method and the ones below it into the SchemaPreservingSolrRDD

thelabdude commented on src/main/java/com/lucidworks/spark/SchemaPreservingSolrRDD.java:
```diff
@@ -0,0 +1,424 @@
+    return new StructField(finalName, st2, true,  Metadata.empty());
+  }
+
+  public static DataType getsqlDataType(String s) {
```
This seems like a useful function for SolrRDD too .. consider moving this method back up into SolrRDD

thelabdude commented on src/main/java/com/lucidworks/spark/SchemaPreservingSolrRDD.java:
```diff
@@ -0,0 +1,424 @@
+    return DataTypes.createArrayType(getsqlDataType(s.split(":")[1]));
+  }
+
+  public JavaRDD<SolrDocument> queryShards(JavaSparkContext jsc, final SolrQuery origQuery, final String splitFieldName, final int splitsPerShard) throws SolrServerException {
```
Please use the @Override annotation when you're overriding a method in SolrRDD

thelabdude commented on src/main/java/com/lucidworks/spark/SolrRelation.java:
```diff
@@ -51,14 +61,23 @@ public SolrRelation(SQLContext sqlContext, Map<String,String> config, DataFrame
     this.sqlContext = sqlContext;
-
+    this.sc =  new JavaSparkContext(sqlContext.sparkContext());
+    if (optionalParam(config, PRESERVE_SCHEMA, "N").equals("Y")) {
```
I'd throw a Boolean.parseBoolean in there too, to allow for Y or True or true as input values

thelabdude commented on src/main/java/com/lucidworks/spark/SolrRelation.java:
```diff
@@ -51,14 +61,23 @@ public SolrRelation(SQLContext sqlContext, Map<String,String> config, DataFrame
     if (splitFieldName != null)
       splitsPerShard = Integer.parseInt(optionalParam(config, SOLR_SPLITS_PER_SHARD_PARAM, "10"));
-    solrRDD = new SolrRDD(zkHost, collection);
+    if(!preserveSchema) {
```
Please use consistent style by putting a space after the if keyword before the opening parens, i.e. 

if (!preserveSchema)

thelabdude commented on src/main/java/com/lucidworks/spark/SolrRelation.java:
```diff
@@ -1,21 +1,29 @@
 import org.apache.spark.api.java.JavaRDD;
 import org.apache.spark.api.java.JavaSparkContext;
 import org.apache.spark.api.java.function.Function;
+import org.apache.spark.mllib.linalg.Matrices;
```
Please remove unused imports (Intellij will do that for you)

thelabdude commented on src/main/java/com/lucidworks/spark/SchemaPreservingSolrRDD.java:
```diff
@@ -0,0 +1,424 @@
+import org.apache.spark.sql.Row;
+
+/**
+ * Created by ganeshkumar on 9/29/15.
```
Please provide a useful description for this class or remove the JavaDoc comment

thelabdude commented on src/main/java/com/lucidworks/spark/SchemaPreservingSolrRDD.java:
```diff
@@ -0,0 +1,424 @@
+    JavaRDD<Row> rows = docs.map(new Function<SolrDocument, Row>() {
+      @Override
+      public Row call(SolrDocument solrDocument) throws Exception {
+        Row ret =  readData(solrDocument, getSolrClient(zkHost) ,  schema, collection);
```
Since you're accessing an expensive object to allocate (SolrClient), you should wrap this in a foreach partition ... see foreachPartition in indexDocs in SolrSupport as an example.

thelabdude commented on src/test/java/com/lucidworks/spark/SolrRelationTest.java:
```diff
@@ -92,6 +106,51 @@ public void testFilterSupport() throws Exception {
   }
 
+  @Test
+  public void testNestedDataFrames() throws Exception {
```
From what I understand, the preserve schema feature is to support storing ML models in Solr? if so, can you add a unit test that builds up a ML model, maybe something simple like a logistic regression and then store it into Solr and retrieve it back?

ganeshk7 commented:
@thelabdude I have made the necessary changes as suggested. For minimizing the use of the expensive object (SolrClient), I have added couple of spark transformations and actions to collect the results in a hashmap instead of recursive solr queries. The hashmap is used in the recursive reads. I have also added couple of unit tests for storing Logistic Regression and Naive Bayes models in Solr. Please review when you get a chance. Thanks

thelabdude commented on src/main/java/com/lucidworks/spark/SolrRelation.java:
```diff
@@ -51,14 +51,23 @@ public SolrRelation(SQLContext sqlContext, Map<String,String> config, DataFrame
     this.sqlContext = sqlContext;
-
+    this.sc =  new JavaSparkContext(sqlContext.sparkContext());
+    if (optionalParam(config, PRESERVE_SCHEMA, "N").equals("Y") || Boolean.parseBoolean(optionalParam(config, PRESERVE_SCHEMA, "False"))) {
```
This should be:

```
String preserveSchema = optionalParam(config, PRESERVE_SCHEMA, "N");
if ("Y".equals(preserveSchema) || Boolean.parseBoolean(preserveSchema)) {
...
```

ganeshk7 commented:
@thelabdude Done Tim.

thelabdude commented on src/main/java/com/lucidworks/spark/SchemaPreservingSolrRDD.java:
```diff
@@ -0,0 +1,537 @@
+      if (x3.length > x.size() && x1.get(x3[x.size()]+"_s") == null && !st.fields()[x.size()].dataType().typeName().equals("struct")){
+        x.add(null);
+      }
+      if ((x2[i].toString().substring(x2[i].toString().length() - 2, x2[i].toString().length()).equals("_s") && !x2[i].toString().startsWith("__lw") && !x2[i].toString().startsWith("links"))) {
```
whoa! that's a lot of toString()'s in there! and many seem repeated ... consider just doing it once and using throughout ... 

Also, what's the story with timestamp and date?

Lastly, if I was writing this, I'd probably use reflection to map type to a convertToX method and then cache that ... but that's mostly stylistic

ganeshk7 commented:
@thelabdude : Sorry about that. Added date and timestamp conversion mechanisms. Also, changed the toString() repetitions to refer to a single variable.

ganeshk7 commented:
Master is up to date with this branch.
